src/v/rpc/transport.cc

```diff
@@ -134,7 +134,7 @@ transport::send(netbuf b, rpc::client_opts opts) {
 
 ss::future<result<std::unique_ptr<streaming_context>>>
```

Comment:
> An important RPC invariant is that resource_units supplied by the caller
must not outlive the request (otherwise it is possible that corresponding

do you mean that they _must outlive_, rather than _not outlive_?

Reply:
I think the sentence is correct. if i understand this correctly we should always return units to semaphore before we return control back to the caller.

Reply:
Michal is correct. Detailed explanation: what we want in the end is that the consensus semaphore must outlive resource_units. A gate in the consensus object already ensures that all RPC requests finish before the consensus object (and therefore the semaphore) is destroyed. So to ensure that the consensus semaphore outlives resource_units, it is enough to guarantee that resource_units *do not* outlive the request.

Reply:
> it is enough to guarantee that resource_units do not outlive the request.

makes complete sense, thanks. I didn't make the connection with the gate in consensus and the draining of requests leading to the implication that the lifetimes had this relationship.